### PR TITLE
Fix memory leak in _DraggableScrollableSheetState

### LINKED ISSUE #133705: _DraggableScrollableSheetState does not dispose extent.
PR that marks the leak: https://github.com/flutter/flutter/pull/133693

Steps to repro:

1. Remove notDisposedAllowList added by the PR
2. Run the updated test library with 'flutter run '
3. Observe the test failure

I tried to invoke ` _extent.dispose()` in `dispose`, but it caused some tests failing: 

```
06:50 +8815 ~22: /b/s/w/ir/x/w/flutter/packages/flutter/test/material/scaffold_test.dart: Floating action button shrinks when bottom sheet becomes dominant                                            
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown while finalizing the widget tree:
A ValueNotifier<double> was used after being disposed.
Once you have called dispose() on a ValueNotifier<double>, it can no longer be used.

When the exception was thrown, this was the stack:
#0      ChangeNotifier.debugAssertNotDisposed.<anonymous closure> (package:flutter/src/foundation/change_notifier.dart:179:9)
#1      ChangeNotifier.debugAssertNotDisposed (package:flutter/src/foundation/change_notifier.dart:186:6)
#2      ChangeNotifier.dispose (package:flutter/src/foundation/change_notifier.dart:375:27)
#3      _DraggableSheetExtent.dispose (package:flutter/src/widgets/draggable_scrollable_sheet.dart:609:18)
#4      _DraggableScrollableSheetState.dispose (package:flutter/src/widgets/draggable_scrollable_sheet.dart:723:13)
#5      StatefulElement.unmount (package:flutter/src/widgets/framework.dart:5669:11)
#6      _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2056:13)
#7      _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2054:7)
#8      ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5512:14)

```

### LINKED ISSUE #134209: Memory Leak: _DraggableScrollableSheetState doesn't dispose _DraggableSheetExtent
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

`_DraggableScrollableSheetState` doesn't dispose `_DraggableSheetExtent` when `DraggableScrollableController` is not provided.

1. `_DraggableScrollableSheetState` creates `_DraggableSheetExtent`:
https://github.com/flutter/flutter/blob/3744050be395f478808e809c423e752c6ef82048/packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart#L640-L658
https://github.com/flutter/flutter/blob/3744050be395f478808e809c423e752c6ef82048/packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart#L726-L743

2. `_DraggableScrollableSheetState` delegates disposal of `_DraggableSheetExtent` to the `DraggableScrollableController`. But if `DraggableScrollableController` is not provided by the widget, the `_DraggableSheetExtent` won't be disposed:
https://github.com/flutter/flutter/blob/3744050be395f478808e809c423e752c6ef82048/packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart#L719-L724

### Expected results

`_DraggableScrollableSheetState` disposes `_DraggableSheetExtent` when `DraggableScrollableController` is not provided.

### Actual results

`_DraggableScrollableSheetState` doesn't dispose `_DraggableSheetExtent` when `DraggableScrollableController` is not provided.

### Code sample

<details><summary>Test which reveals the leak:</summary>

```dart
testWidgetsWithLeakTracking('_DraggableScrollableSheetState leak', (WidgetTester tester) async {
  await tester.pumpWidget(
    Directionality(
      textDirection: TextDirection.ltr,
      child: MediaQuery(
        data: const MediaQueryData(),
        child: Align(
          alignment: Alignment.bottomCenter,
          child: DraggableScrollableSheet(
            builder: (BuildContext context, ScrollController scrollController) {
              return const Center();
            },
          ),
        ),
      ),
    ),
  );
}, leakTrackingTestConfig: LeakTrackingTestConfig.debugNotDisposed());
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel master, 3.14.0-14.0.pre.142, on macOS 13.0.1 22A400 darwin-arm64, locale en-GB)
    • Flutter version 3.14.0-14.0.pre.142 on channel master at /Users/ksokolovskyi/dev/flutter_master
    • Upstream repository git@github.com:ksokolovskyi/flutter.git
    • FLUTTER_GIT_URL = git@github.com:ksokolovskyi/flutter.git
    • Framework revision 3744050be3 (2 hours ago), 2023-09-07 05:27:18 -0400
    • Engine revision 5a45ecd24a
    • Dart version 3.2.0 (build 3.2.0-134.0.dev)
    • DevTools version 2.27.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /Users/ksokolovskyi/Library/Android/sdk
    • Platform android-33, build-tools 33.0.1
    • ANDROID_HOME = /Users/ksokolovskyi/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Community Edition (version 2023.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.81.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.72.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.0.1 22A400 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 116.0.5845.140

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by ksokolovskyi**
### Description
- Fixes a memory leak in `_DraggableScrollableSheetState` where `_DraggableSheetExtent` wasn't disposed when `DraggableScrollableController` wasn't provided.

### Related Issues
- Fixes: #134209;
- Fixes: #133705;
- Removes leak mark: #133693.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

### CODE REVIEW

**Comment by ksokolovskyi** [comment]
cc @polina-c

**Review by polina-c — APPROVED** [review]
One comment then LGTM

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart [hunk 1]
```diff
@@ -718,7 +718,11 @@ class _DraggableScrollableSheetState extends State<DraggableScrollableSheet> {
 
   @override
   void dispose() {
-    widget.controller?._detach(disposeExtent: true);
+    if (widget.controller != null) {
```
Comment (polina-c) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:721]: Can you revert it o make more readable:

`if (widget.controller == null) {`

Comment (ksokolovskyi) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart]: @polina-c thanks a lot for the review!
Added the fix.

**Pull request merged**